File: Algorithms/CPP/1367 Linked List in Binary Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (head == NULL || root == NULL) {
            return false;
        }

        return explorePath(head, root) || isSubPath(head, root->left) || isSubPath(head, root->right);
    }

    bool explorePath(ListNode* node, TreeNode* treeNode) {
        if (node == NULL) {
            return true;
        }

        if (treeNode == NULL) {
            return false;
        }

        if (treeNode->val != node->val) {
            return false;
        }
        
        return explorePath(node->next, treeNode->left) || explorePath(node->next, treeNode->right);
    }
};

File: Algorithms/Python/1367 Linked List in Binary Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# double recursion
class Solution:
    def isSubPath(self, head: ListNode, root: TreeNode) -> bool:
        if head is None or root is None:
            return False

        return self.explorePath(head, root) or self.isSubPath(head, root.left) or self.isSubPath(head, root.right)

    def explorePath(self, node, treeNode):
        if node is None:
            return True

        if treeNode is None:
            return False

        if node.val != treeNode.val:
            return False

        return self.explorePath(node.next, treeNode.left) or self.explorePath(node.next, treeNode.right)
